nodeJS/express/views.md

```diff
@@ -0,0 +1,180 @@
+
+When you hit the `/` route, `res.render("index", { message: "EJS rocks!" });`  is the line that sends back the response. Since we've already defined the `views` and `view engine` app properties, the first argument of `res.render` is programmed to look for "a template called index in the specified folder", while the second argument is an object of variables that are to be made available to that specific template.
+
+### Reusable templates
+
+You may want to include webpage components that are shared across different pages, such as a sidebar or a header. To insert such components into your pages, we make use of the `include` command. This requires the name of the file to be inserted, and optionally an object of data you wish to pass.
+
+Say you have the following navbar component called `"navbar.ejs"`:
+
+```html
+<!-- navbar.ejs -->
+<nav>
+  <ul>
+    <% for (let i = 0; i < links.length; i++) { %>
+      <li>
+        <a href="<%= links[i].href %>">
+          <span> <%= links[i].text %> </span>
+        </a>
+      </li>
+    <% } %>
+  </ul>
+</nav>
+```
+
+You can insert this component into another EJS file like so:
+
+```html
+<!-- index.ejs -->
+<html>
+  <head>
+    <title>Homepage</title>
+    <body>
+      <%- include('navbar', {links: links}) %>
+    </body>
+  </head>
+</html>
+```
+
+This can be used to include headers and footers in all of your pages, for example.
+
+Note that the navbar expects a `links` value. To pass this data into the navbar, you can pass it when rendering `index.ejs` which contains the navbar. Modify `app.js` such that a `links` object is defined and passed into the `render` function in the `"/"` route handler:
+
+```javascript
+// app.js
+const links = [
+  { href: "/", text: "Home" },
+  { href: "about", text: "About" },
+];
+
+app.get("/", (req, res) => {
+  res.render("index", { links: links });
+});
+```
+
+Here's another example of how to use `includes` to dynamically render a list of variables:
+
+```html
+<ul>
+  <% users.map( user => { %> <%- include('user/show', {user: user}) %> <% }) %>
+</ul>
+```
+
+<div class="lesson-note lesson-note--tip" markdown="1">
+
+#### Directories within the views folder
+
+We can have nested directories of EJS template files within the views. For example, to render the template file `./views/user/show.ejs`, we'll need to provide the relative path like so:
+
+```javascript
+// in res.render
+res.render("user/show")
+
+// in include
+include("user/show")
+```
+
+Note the use of the raw output tag `<%-` with the `include` which is used to avoid double-escaping the HTML output.
+
+### Assignment
+
+<div class="lesson-content__panel" markdown="1">
+
+1. Read through the [Express resource on template engines](https://expressjs.com/en/guide/using-template-engines.html). The resource uses Pug for the examples which has a different syntax, however the information should still be a useful supplement to this lesson.
```

Comment:
Can we also include an assignment item for the EJS docs?

Reply:
Thanks for the suggestion, I've just added EJS Docs to Assignment! I thought it wouldn't be necessary because there is already a link to the docs in the introduction, but perhaps learners will benefit from a clear instruction to read it.